Updates to the authorization documentation

whitehatguy opened the pull request:
Updates guidance, including a specification document and incorporates the OpenID Connect usage guidance.

whitehatguy commented:
Dennis had made some comments on the OpenID pull request (https://github.com/cerner/fhir.cerner.com/pull/71) -- I'll go ahead and add those example links before we merge this pull request.

MaxPhilips commented:
Looks like the error URNs are removed with these changes?  Are they hosted somewhere else?  The authorization server "An Error has Occurred" page gives a link to http://fhir.cerner.com/dstu2/authorization/#error-codes so we should update that link if this information is moving

whitehatguy commented:
@dennispatterson , I've updated the OpenID Connect documents with examples now.

whitehatguy commented:
@MaxPhilips - I've stated a spike on the topic of having an error listing.  I think the original intent was to have a link to Cerner's specifications for developers, but it ended up becoming a link to an error listing (which isn't helpful at all - it just regurgitates what we've already told them.)  We're going to settle this internally, but at this point, I don't plan on manually maintaining an ever-changing list of error codes in this documentation.

https://jira.cerner.com/browse/CLOUDAUTHZ-802

MaxPhilips commented on content/dstu2/authorization-specification.md:
```diff
@@ -0,0 +1,1029 @@
+
+Non-normative conformance example as JSON:
+
+<pre class="terminal">
```
Can you reformat this as
```
{
  "resourceType": "Conformance", 
  ...
  "rest": [
    {
      ...
      "security": {
        "extension": [
          {
            "url": "http://FHIR-registry.smarthealthit.org/StructureDefinition/oauth-uris",
            "extension": [
              {
                "url": "token",
                "valueUri": "https://my-server.org/token"
              },
              {
                "url": "authorize",
                "valueUri": "https://my-server.org/authorize"
              },
              {
                "url": "manage",
                "valueUri": "https://my-server.org/manage"
              }
            ]
          }
        ],
      ...
```
to match JSON examples on other pages?

MaxPhilips replied:
https://github.com/cerner/fhir.cerner.com/pull/99/commits/59ab440f6b0251957c8d66a4c9616ffd1280dc37#diff-cc928c50e228a325bdbe6f02c205841fR846

MaxPhilips commented on content/dstu2/authorization-specification.md:
```diff
@@ -0,0 +1,1029 @@
+### A.1. Resource Scope Syntax ###
+
+<pre class="terminal">
+scope-name            = resource-context "/" resource-type "." modification-rights
```
The = signs in these 5 sections currently align to 3 different columns, could you move A1 left by 2 spaces and A3-5 right by 1 space? :)

MaxPhilips replied:
https://github.com/cerner/fhir.cerner.com/pull/99/commits/59ab440f6b0251957c8d66a4c9616ffd1280dc37#diff-cc928c50e228a325bdbe6f02c205841fL969

MaxPhilips commented on content/dstu2/authorization.md:
```diff
@@ -2,195 +2,1061 @@
-   "token_type":"Bearer",
-   "expires_in":570,
-   "refresh_token":"1234-567890ab-cdef"
+<pre class="terminal">
```
As-is this is missing one closing bracket and can you also update the formatting to
```
{
  "resourceType": "Conformance",
  "rest": [
    {
      "security": {
        "extension": [
          {
            "url": "http://fhir-registry.smarthealthit.org/StructureDefinition/oauth-uris",
            "extension": [
              {
                "url": "token",
                "valueUri": "https://example.org/token"
              },
              {
                "url": "authorize",
                "valueUri": "https://example.org/authorize"
              }
            ]
          }
        ]
      }
    }
  ]
}
```
?

MaxPhilips replied:
https://github.com/cerner/fhir.cerner.com/pull/99/commits/59ab440f6b0251957c8d66a4c9616ffd1280dc37#diff-87e607cdcd40ece132eef7d65ad4e986R128

MaxPhilips commented on content/dstu2/authorization.md:
Add registered tm symbol

MaxPhilips commented on content/dstu2/authorization.md:
Deindent other vars to match width? Or indent location to a full tab

MaxPhilips commented on content/dstu2/authorization.md:
Add sup tags

MaxPhilips commented on content/may2015/authorization-specification.md:
```diff
@@ -0,0 +1,1029 @@
+---
```
I assume the same comments apply to the may2015 versions of the files as the dstu2 versions

MaxPhilips commented on content/dstu2/authorization-specification.md:
```diff
@@ -0,0 +1,1029 @@
+title: Authorization | DSTU 2 API
+---
+
+# Cerner FHIR® Service Authorization Specification v1.0 #
```
Add sup tag

MaxPhilips commented on content/dstu2/authorization.md:
```diff
@@ -2,195 +2,1061 @@
 
-### Requesting an access token ###
-Once an authorization code has been acquired, a back-channel **POST** to the authorization server's token endpoint can be made to request an access token. Following the [OAuth 2.0][1] spec, the authorization server accepts a content type of **application/x-www-form-urlencoded** from your client application with the following information.
+The FHIR® Conformance resource can be retrieved by performing
```
sup

dennispatterson commented on content/dstu2/authorization.md:
```diff
@@ -2,195 +2,1061 @@
-Once an authorization code has been acquired, a back-channel **POST** to the authorization server's token endpoint can be made to request an access token. Following the [OAuth 2.0][1] spec, the authorization server accepts a content type of **application/x-www-form-urlencoded** from your client application with the following information.
+The FHIR® Conformance resource can be retrieved by performing
+an HTTP GET against the resource located at the relative path
+of ``../Conformance`` from the FHIR<sup>®</sup> base URL.
```
This makes it sound like the Conformance endpoint is "https://fhir-ehr.sandboxcerner.com/dstu2/Conformance" rather than "https://fhir-ehr.sandboxcerner.com/dstu2/0b8a0111-e8e6-4c26-a91c-5069cbc6b1ca/metadata".  May want to replace "../Conformance" with "./metadata"

whitehatguy replied:
Will correct.

dennispatterson replied:
Matt, I [issued a PR](https://github.com/whitehatguy/fhir.cerner.com/pull/2) to your branch on the 13th to help address the issues that I raised.  Hopefully that'll save you some time.

MaxPhilips replied:
I can do the same (or commit directly to your fork) for the formatting changes I found

dennispatterson replied:
@whitehatguy I see you just replied to another thread below with intentions to make changes, so just making sure you see this thread with the PR I mention and Max's offer.

dennispatterson commented on content/dstu2/authorization.md:
```diff
@@ -2,195 +2,1061 @@
+
+- ``user/Observation.read`` (Read all values from the Observation resource
+  that the user has access to.)
+- ``patient/MedicationOrder.*`` (Read or write medications for a specific user.)
```
user -> patient

dennispatterson commented on content/dstu2/authorization.md:
```diff
@@ -2,195 +2,1061 @@
+The scope 'profile' will additionally request that the OpenID Connect
+token include the claim "profileURL", as defined by the 
+SMART<sup>®</sup> on FHIR<sup>®</sup> authorization framework.  This 
+URL identifiesthe specific FHIR<sup>®</sup> resource URL of the 
```
nit: identifiesthe -> identifies the

dennispatterson commented on content/dstu2/authorization.md:
This looks more like the url for the auth server and not the url for the FHIR server

dennispatterson commented on content/dstu2/authorization.md:
Isn't the element just "patient" and not "patient_id"?

dennispatterson commented on content/dstu2/authorization.md:
I'm guessing many readers aren't going to have access to the uCern wiki.  Just wondering if we should some sort of comment.

dennispatterson commented on content/dstu2/authorization-specification.md:
```diff
@@ -0,0 +1,1029 @@
+
+####  Access Token Request ####
+SMART applications __SHALL__ utilize an access token request as 
+described in [section 4.4.3 of OAuth 2.0](http://tools.ietf.org/html/rfc6749#section-4.1.3).  
```
Display section and link are not consistent

whitehatguy replied:
Will correct.

dennispatterson commented on content/may2015/openid-connect.md:
```diff
@@ -0,0 +1,153 @@
+
+The client application receiving the identity token must validate that 
+the audience (aud) claim matches its own client identifier, per section
+3.1.3.7 of [OpenID Connect Core][OPENID].
```
Link to specific sections in this document?

dennispatterson commented on content/may2015/openid-connect.md:
```diff
@@ -0,0 +1,153 @@
+    "require_request_uri_registration": false,
+    "request_uri_parameter_supported": true,
+    "claims_parameter_supported": false,
+    "jwks_uri": "https:\/\/authorization.sandboxcerner.com\/jwk",
```
Are the extra escapes needed?

whitehatguy replied:
The example came from the actual responses from the server - probably best the examples are in-line with actual output.

dennispatterson commented on content/may2015/openid-connect.md:
```diff
@@ -0,0 +1,153 @@
+---
+title: OpenID Connect | MAY 2015 BALLOT APIa
```
nit: extra 'a' at the end

whitehatguy commented on content/dstu2/authorization-specification.md:
```diff
@@ -0,0 +1,1029 @@
+---
+title: Authorization | DSTU 2 API
```
No.  Will update.

whitehatguy commented on content/dstu2/authorization.md:
You're correct.  I'll update the example.

whitehatguy commented on content/dstu2/authorization.md:
You're right, it is just patient, and encounter would also be valid.

daliboz commented:
@whitehatguy  I think you accidentally moved all these changes to may 2015 in that last commit, including the dstu2 specific links and all the documentation that only applies to DSTU 2

MaxPhilips commented on content/dstu2/authorization-specification.md:
```diff
@@ -0,0 +1,1029 @@
+
+### Contextless Flow ###
+
+<pre class="terminal">                                                                                                                            
```
https://github.com/cerner/fhir.cerner.com/pull/99/commits/59ab440f6b0251957c8d66a4c9616ffd1280dc37#diff-cc928c50e228a325bdbe6f02c205841fR117

MaxPhilips commented on content/dstu2/authorization-specification.md:
```diff
@@ -0,0 +1,1029 @@
+                            |               |            |              |
+                            +---------------+            +--------------+
+
+
```
https://github.com/cerner/fhir.cerner.com/pull/99/commits/59ab440f6b0251957c8d66a4c9616ffd1280dc37#diff-cc928c50e228a325bdbe6f02c205841fR144

MaxPhilips commented on content/dstu2/authorization-specification.md:
```diff
@@ -0,0 +1,1029 @@
+
+### EHR Launch Flow ###
+
+<pre class="terminal">                                                   
```
https://github.com/cerner/fhir.cerner.com/pull/99/commits/59ab440f6b0251957c8d66a4c9616ffd1280dc37#diff-cc928c50e228a325bdbe6f02c205841fR261

MaxPhilips commented on content/dstu2/authorization-specification.md:
```diff
@@ -0,0 +1,1029 @@
+
+### SMART Application Launch Flow ###
+
+<pre class="terminal">                         
```
https://github.com/cerner/fhir.cerner.com/pull/99/commits/59ab440f6b0251957c8d66a4c9616ffd1280dc37#diff-cc928c50e228a325bdbe6f02c205841fR425

MaxPhilips commented on content/dstu2/authorization.md:
https://github.com/cerner/fhir.cerner.com/pull/99/commits/d2f601c4a34782882aa9745140e40ab8ce7de7b6#diff-87e607cdcd40ece132eef7d65ad4e986R456

MaxPhilips commented:
Pushed 2 commits addressing my most of my comments.  This comment: "There's no step 12 in this flow" still applies to content/dstu2/authorization-specification.md L465

dennispatterson commented:
The May version of the OpenId Connect page can be pulled since we're not linking to it anywhere.

dennispatterson commented on content/dstu2/authorization-specification.md:
```diff
@@ -0,0 +1,1029 @@
+	how the SMART application is provided the URL for the FHIR<sup>®</sup> 
+	server is not defined by this specification.
+3.  The FHIR<sup>®</sup> server returns the conformance statement, 
+    which provides the needed endpoints for steps 8 and 12.
```
@MaxPhilips if you want to update this to steps 4 and 8, that'll correct the issue.  (If you also remove the May open-id connect page, that may resolve our outstanding issues.)

MaxPhilips commented on Gemfile.lock:
```diff
@@ -17,6 +17,7 @@ GEM
     ethon (0.9.1)
       ffi (>= 1.3.0)
     ffi (1.9.14)
+    ffi (1.9.14-x64-mingw32)
```
I didn't notice this before but I'm guessing these -x64-mingw32 gem versions showed up due to bundle installing on Windows?  We likely do not need to commit these Gemfile.lock changes

whitehatguy replied:
I'll fix.  Is there a way to avoid this in the future by chance?  Seems like the gemfile gets modified as a result of running doc generation - should it just be in the .gitignore ?

MaxPhilips replied:
Not needing to commit these specific changes is all I meant.  Off the top of my head, I'm not sure if committing Gemfile.lock is necessary for the gh-pages deployment - I assume it is.  I'll look into this and see if we could .gitignore it.

whitehatguy commented:
@dennispatterson - I'll proceed with pulling the OpenIDConnect doc from the may folder.

whitehatguy commented:
Corrected the references to "steps" in the diagrams from conformance discovery, and reverted Gemfile.lock.  I think it's caught up now to all comments and pull requests.

MaxPhilips commented:
+1, would you please squash the commits before merging?

whitehatguy commented:
Sure, I'll submit a new pull request with the squashed updates.
